### PR TITLE
[SPARK-8755][Streaming]Login user before reading the checkpoint file in hdfs.

### PR DESCRIPTION
**Pull request by SaintBacchus**
If the user set `spark.yarn.principal` and `spark.yarn.keytab` , he does not need `kinit` in the client machine.
But when the application was recorved from checkpoint file, it had to `kinit`, because:
The checkpoint did not use this configurations before it use a DFSClient to fetch the ckeckpoint file.

But there is a small problem, the `UserGroupInformation.loginUserFromKeytab` will be called twice in checkpoint application. Ignored it in this PR.

[Jira Address](https://issues.apache.org/jira/browse/SPARK-8755)

### CODE REVIEW

**Comment by tdas** [comment]
@harishreedharan @srowen Can you take a look at this?

**Comment by srowen** [comment]
I'm not a YARN person so phoning @sryza

File: streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala [hunk 1]
```diff
@@ -324,6 +326,17 @@ object CheckpointReader extends Logging {
     }
     None
   }
+
+  /**
+   * If using the keytab to login user in security mode, this checkpoint application has to login
+   * before reading files from hdfs.
+   */
+  private def loginUserIfConfExsis(conf: SparkConf): Unit = {
```
Comment (andrewor14) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: Exists

**Comment by andrewor14** [comment]
Looks good. @harishreedharan should confirm once he's back from vacation.

File: streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala [hunk 2]
```diff
@@ -324,6 +326,17 @@ object CheckpointReader extends Logging {
     }
     None
   }
+
+  /**
+   * If using the keytab to login user in security mode, this checkpoint application has to login
+   * before reading files from hdfs.
```
Comment (andrewor14) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: maybe add that this is only for YARN mode.

Comment (harishreedharan) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: I will look at this today.

On Monday, July 6, 2015, andrewor14 <notifications@github.com> wrote:

> In streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala
> <https://github.com/apache/spark/pull/7158#discussion_r33969975>:
>
> > @@ -324,6 +326,17 @@ object CheckpointReader extends Logging {
> >      }
> >      None
> >    }
> > +
> > +  /**
> > +   * If using the keytab to login user in security mode, this checkpoint application has to login
> > +   * before reading files from hdfs.
>
> maybe add that this is only for YARN mode.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7158/files#r33969975>.
>


-- 

Thanks,
Hari

**Comment by harishreedharan** [comment]
From what I can see this is not a real issue in Cluster mode. The re-started app will actually be able to read from HDFS using the credentials created for it by Client.scala here: https://github.com/apache/spark/blob/master/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala#L795

In Client mode though, this is an issue - but that is actually a bug, since in Client mode - Client.scala does not do a login because of what presumably is a bug. I am opening a PR for that soon.

**Comment by SaintBacchus** [comment]
@harishreedharan this is an issue only in Client mode. 
Will your PR cover this issue?

**Comment by harishreedharan** [comment]
Yep. I think so. I will post it in a couple hours. You can test it out.

On Monday, July 6, 2015, huangzhaowei <notifications@github.com> wrote:

> @harishreedharan <https://github.com/harishreedharan> this is an issue
> only in Client mode.
> Will your PR cover this issue?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7158#issuecomment-119043741>.
>


-- 

Thanks,
Hari

**Comment by SaintBacchus** [comment]
OK, I close this PR now.

**Comment by SaintBacchus** [comment]
@tgravescs I report this issue. Can you also take a look?

**Comment by tgravescs** [comment]
so this ties with https://github.com/apache/spark/pull/7255. 
Why is that not fixing the issue?  You should already be logged in before the first renewal is needed and if the application master code to login and propagate the credentials is working it should again already be logged in.  

It seems more like something isn't being propagated properly so I would like to understand exactly what is going on.  You mention in the description "The checkpoint did not use this configurations before it use a DFSClient"   Can you clarify this?

**Comment by SaintBacchus** [comment]
@tgravescs the checkpoint had done this three things:
1.Read the checkpoint file
2.Deserialize checkpoint file and get the properties
3.Initialize the `SparkContext`
The issue was in Step One.  The code in `Client.scala` will run in Step Three.

**Comment by tgravescs** [comment]
ah ok, if that is the order it makes sense.  Let me look at the code a bit more.

**Comment by harishreedharan** [comment]
From what I remember, the code in Client.scala executes first since that is
called from SparkSubmit. The only thing I am not sure of is whether the
tokens are getting added before or after the user class is started. If it
is getting after user class is started then that is likely the issue. I
will look at this in some time.

On Tuesday, July 7, 2015, Tom Graves <notifications@github.com> wrote:

> ah ok, if that is the order it makes sense. Let me look at the code a bit
> more.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7158#issuecomment-119222348>.
>


-- 

Thanks,
Hari

**Comment by harishreedharan** [comment]
OK, so I looked into this one. The issue is that in Client mode - the user class is started first, which in turn, starts the `YarnClientSchedulerBackend` which creates the `Client` instance. Since the user class would actually end up reading the checkpoint first, this seems like a valid issue. 

I don't see an alternate solution without hacking `SparkSubmit` to do a login.

**Comment by tgravescs** [comment]
So is this only an issue if the user hasn't kinit'd before submission then?

**Comment by harishreedharan** [comment]
Yep. If the user kinits before running the app, then this should not be an issue. One alternate (unclean) solution is to expose the login method from the `Client` in the companion object (since we don't want multiple Client instances) and call `setupCredentials` from `SparkSubmit`. That would fix this issue and the one in #7255.

**Comment by tgravescs** [comment]
Personally I don't consider this a big issue then.  I think user should have to kinit before submission. Its a nice thing that we login for them if they specify the keytab and principal but I never saw that as a requirement or guaranteed feature.  It is kind of bad that it is not consistent between modes though.  

Putting this down inside checkpoint seems kind of hacky and one off to me.  someone could add something else before this that would access hdfs and then it would break again.  If we are going to make this an actual supported feature we should pull it up higher and make sure its one of the first things that happen.

**Comment by harishreedharan** [comment]
@tgravescs I agree. Let me update the patch in #7255 to do the login from `SparkSubmit` - it is not really a "clean" way of doing it, but it is will work even if the user code accesses HDFS directly.

**Comment by andrewor14** [comment]
This patch is superseded by #7255 right? Should we close it?

**Pull request closed without merging**